File: docs/source/src/python/gp/trees.py
# --8<-- [start:build_tree]
import radiate as rd

codec = rd.TreeCodec(
    shape=(2, 1),  # 2 inputs, 1 output
    min_depth=3,  # starting depth of the tree (default 3)
    max_size=30,  # max number of nodes (default 30)
    vertex=[rd.Op.add(), rd.Op.sub(), rd.Op.mul(), rd.Op.div()],
    leaf=[rd.Op.var(0), rd.Op.var(1)],
)

# decode a (randomly initialized) genotype into a Tree
tree = codec.decode(codec.encode())

result = tree.eval([1.0, 2.0])  # list[float], one value per output
# --8<-- [end:build_tree]

# --8<-- [start:tree_codec_detailed]
import radiate as rd

# Create a tree codec with a starting (minimum) depth of 3
codec = rd.TreeCodec(
    # Shape is only necessary for multi-rooted trees,
    # it specifies the number of trees and the
    # number of inputs to each tree.
    # For single rooted trees this will be ignored.
    #
    # Default: shape=(1, 1) - single rooted tree with 1 input variable
    shape=(2, 1),
    # The minimum depth of the tree. This is the depth
    # of the initial tree that will be produced by
    # the codec - evolution will be able
    # to increase the total size of the tree (without regards to depth) up to max_size.
    #
    # Default: min_depth=3
    min_depth=3,
    # The maximum size of the tree. This is the maximum number of nodes that the tree can have.
    max_size=30,
    # The options for the root node of the tree. The root is the last node in the tree
    # to be evaluated and is the node that produces the final output of the tree.
    # This is optional and defaults to a single add operator.
    root=rd.Op.add(),
    vertex=[rd.Op.add(), rd.Op.sub(), rd.Op.mul()],
    leaf=[rd.Op.var(0), rd.Op.var(1)],
)

genotype = codec.encode()
tree = codec.decode(genotype)
# --8<-- [end:tree_codec_detailed]

# --8<-- [start:hoist_mutator]
import radiate as rd

mutator = rd.HoistMutator(rate=0.1)
mutator = rd.Mutate.hoist(0.1)  # Using the dsl syntax for mutators
# --8<-- [end:hoist_mutator]

# --8<-- [start:tree_crossover]
import radiate as rd

crossover = rd.TreeCrossover(rate=0.1)
crossover = rd.Cross.tree(0.1)  # Using the dsl syntax for crossover operators
# --8<-- [end:tree_crossover]

File: docs/source/src/rust/gp/trees.rs
use radiate::prelude::*;

fn main() {
    // --8<-- [start:build_tree]
    // create a simple tree:
    //              42
    //           /  |   \
    //          1   2    3
    //             / \
    //            3   4
    let tree: Tree<i32> = Tree::new(
        TreeNode::new(42)
            .attach(TreeNode::new(1))
            .attach(
                TreeNode::new(2)
                    .attach(TreeNode::new(3))
                    .attach(TreeNode::new(4)),
            )
            .attach(TreeNode::new(3)),
    );

    // The tree can be evaluated with a function that takes a vector of inputs
    // This creates a `Tree` that looks like:
    //      +
    //    /   \
    //   *     +
    //  / \   / \
    // 2  3  2   x
    //
    // Where `x` is the first variable in the input.
    // This can also be thought of (and is functionally equivalent) as:
    //
    // f(x) = (2 * 3) + (2 + x)
    //
    let root = TreeNode::new(Op::add())
        .attach(
            TreeNode::new(Op::mul())
                .attach(TreeNode::new(Op::constant(2.0)))
                .attach(TreeNode::new(Op::constant(3.0))),
        )
        .attach(
            TreeNode::new(Op::add())
                .attach(TreeNode::new(Op::constant(2.0)))
                .attach(TreeNode::new(Op::var(0))),
        );

    // And the result of evaluating this tree with an input of `1` would be:
    let result = root.eval(&vec![1_f32]);
    assert_eq!(result, 9.0);
    // --8<-- [end:build_tree]

    // --8<-- [start:tree_codec]
    let store = vec![
        (NodeType::Root, vec![Op::add(), Op::sub()]),
        (NodeType::Vertex, vec![Op::add(), Op::sub(), Op::mul()]),
        (NodeType::Leaf, vec![Op::constant(1.0), Op::constant(2.0)]),
    ];

    // Create a single rooted tree codec with a starting (minimum) depth of 3
    let codec = TreeCodec::single(3, store.clone());
    let genotype: Genotype<TreeChromosome<Op<f32>>> = codec.encode();
    let tree: Tree<Op<f32>> = codec.decode(&genotype);

    // Create a multi-rooted tree codec with a starting (minimum) depth of 3 and 2 trees
    let codec = TreeCodec::multi_root(3, 2, store);
    let genotype: Genotype<TreeChromosome<Op<f32>>> = codec.encode();
    // multi-rooted codec decodes to a Vec of Trees
    // one for each root in the genotype
    let trees: Vec<Tree<Op<f32>>> = codec.decode(&genotype);
    // --8<-- [end:tree_codec]

    // --8<-- [start:hoist_mutator]
    let mutator = HoistMutator::new(0.1);
    // --8<-- [end:hoist_mutator]

    // --8<-- [start:tree_crossover]
    let mutator = TreeCrossover::new(0.1);
    // --8<-- [end:tree_crossover]
}
